File: modules/strategy-core/src/main/java/com/cryptostrategy/platform/strategy/internal/parameter/StrategyParameterValidator.java
package com.cryptostrategy.platform.strategy.internal.parameter;

import com.cryptostrategy.platform.strategy.api.error.StrategyErrorCode;
import com.cryptostrategy.platform.strategy.api.error.StrategyException;
import com.cryptostrategy.platform.strategy.api.model.parameter.CrossParameterConstraint;
import com.cryptostrategy.platform.strategy.api.model.parameter.ParameterDefinition;
import com.cryptostrategy.platform.strategy.api.model.parameter.ParameterType;
import com.cryptostrategy.platform.strategy.api.model.parameter.StrategyParameterSchema;
import com.cryptostrategy.platform.strategy.api.model.parameter.StrategyParameterSet;
import com.cryptostrategy.platform.strategy.api.model.parameter.StrategyParameterValue;
import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Map;

public final class StrategyParameterValidator {
    public StrategyParameterSet resolve(StrategyParameterSchema schema, Map<String, StrategyParameterValue> supplied) {
        Map<String, ParameterDefinition> definitions = new HashMap<>();
        schema.definitions().forEach(definition -> definitions.put(definition.name(), definition));
        for (String name : supplied.keySet()) if (!definitions.containsKey(name)) fail("Unknown parameter: " + name);
        Map<String, StrategyParameterValue> resolved = new HashMap<>();
        for (ParameterDefinition definition : schema.definitions()) {
            StrategyParameterValue value = supplied.get(definition.name());
            if (value == null) value = definition.defaultValue().orElse(null);
            if (value == null && definition.required()) fail("Missing parameter: " + definition.name());
            if (value == null) continue;
            if (value.type() != definition.type()) fail("Wrong type: " + definition.name());
            validateValue(definition, value); resolved.put(definition.name(), value);
        }
        StrategyParameterSet result = StrategyParameterSet.of(resolved);
        for (CrossParameterConstraint constraint : schema.constraints()) {
            BigDecimal lower = numeric(result.require(constraint.lowerParameter()));
            BigDecimal upper = numeric(result.require(constraint.upperParameter()));
            if (lower.compareTo(upper) >= 0) fail(constraint.lowerParameter() + " must be less than " + constraint.upperParameter());
        }
        return result;
    }
    private static void validateValue(ParameterDefinition definition, StrategyParameterValue value) {
        if (definition.type() == ParameterType.ENUM && !definition.allowedValues().contains(value.canonicalText())) fail("Value not allowed: " + definition.name());
        if (definition.type() == ParameterType.INTEGER || definition.type() == ParameterType.DECIMAL) {
            BigDecimal number = numeric(value);
            if (definition.minimum().isPresent() && number.compareTo(definition.minimum().get()) < 0) fail("Below minimum: " + definition.name());
            if (definition.maximum().isPresent() && number.compareTo(definition.maximum().get()) > 0) fail("Above maximum: " + definition.name());
        }
    }
    private static BigDecimal numeric(StrategyParameterValue value) {
        if (value instanceof StrategyParameterValue.IntegerValue integer) return BigDecimal.valueOf(integer.value());
        if (value instanceof StrategyParameterValue.DecimalValue decimal) return decimal.value();
        fail("Expected numeric parameter"); return BigDecimal.ZERO;
    }
    private static void fail(String message) { throw new StrategyException(StrategyErrorCode.INVALID_PARAMETERS, message); }
}

File: apps/web/src/features/strategy/state/strategy-parameter-validator.ts
import type { StrategyDescriptor } from "../model/strategy";
export type ValidationIssues = Readonly<Record<string, string>>;
const decimal = /^-?(0|[1-9][0-9]*)(\.[0-9]+)?$/;
function decimalParts(value: string, places: number) {
  const negative = value.startsWith("-");
  const [whole, fraction = ""] = value.replace("-", "").split(".");
  const result =
    BigInt(whole) * 10n ** BigInt(places) + BigInt(fraction.padEnd(places, "0") || "0");
  return negative ? -result : result;
}
function compareDecimal(a: string, b: string) {
  const places = Math.max(a.split(".")[1]?.length ?? 0, b.split(".")[1]?.length ?? 0);
  const left = decimalParts(a, places),
    right = decimalParts(b, places);
  return left < right ? -1 : left > right ? 1 : 0;
}
export function validateStrategyParameters(
  descriptor: StrategyDescriptor,
  values: Readonly<Record<string, string>>
): ValidationIssues {
  const issues: Record<string, string> = {};
  for (const field of descriptor.parameters) {
    const value = values[field.name] ?? "";
    if (field.required && !value) {
      issues[field.name] = "Required.";
      continue;
    }
    if (!value) continue;
    if (["INTEGER", "DECIMAL"].includes(field.type) && !decimal.test(value))
      issues[field.name] = "Enter a valid number.";
    else if (field.type === "INTEGER" && !/^-?(0|[1-9][0-9]*)$/.test(value))
      issues[field.name] = "Must be a whole number.";
    else if (field.type === "BOOLEAN" && !["true", "false"].includes(value))
      issues[field.name] = "Must be true or false.";
    else if (field.type === "ENUM" && !field.allowedValues.includes(value))
      issues[field.name] = "Unsupported value.";
    else if (field.minimum && compareDecimal(value, field.minimum) < 0)
      issues[field.name] = `Minimum value: ${field.minimum}.`;
    else if (field.maximum && compareDecimal(value, field.maximum) > 0)
      issues[field.name] = `Maximum value: ${field.maximum}.`;
  }
  for (const rule of descriptor.constraints) {
    const lower = values[rule.lowerParameter],
      upper = values[rule.upperParameter];
    if (
      lower &&
      upper &&
      decimal.test(lower) &&
      decimal.test(upper) &&
      compareDecimal(lower, upper) >= 0
    )
      issues[rule.upperParameter] = `Must be greater than ${rule.lowerParameter}.`;
  }
  return issues;
}
